mpptChgD/mpptChgDconfig.txt
# mpptChgD configuration file
#
# Configure the following optional functions:
#   1. Auto shutdown enable
#   2. TCP Access
#   3. Charger register value Logging
#   4. Charger configuration parameters
#   5. Watchdog enable
#
# All options shown below. Uncomment to enable. Configuration items have the form
# <ITEM>=<VALUE> where <VALUE> may be an integer or string value.  Items that are
# used to enable functionality (such as SHUTDOWN below) may either be commented out
# or set to 0 to disable.  Uncomment them and set the value to 1 to enable.
#
# Comments are indicated by a '#' or ';' character in the first position.
#

# Automatic Low Battery shutdown.  Uncomment the following line to enable a controlled
# OS shutdown upon detection of an imminent power down due to low battery.
SHUTDOWN=1

# Remote TCP access.  Uncomment and set the TCP_PORT to a non-zero number to enable
# TCP access to the daemon.
TCP_PORT=23000
# Maximum number of simultaneous TCP connections (default is 1 connection, maximum 4)
TCP_MAX=4

# Logging parameters
#
# Configure the logfile
LOG_FILE=/home/pi/mpptChgDlog.txt
#
# Number of seconds between log entries
LOG_DELAY=60
#
# Log items.  Uncomment charger register values you wish to log.  Important to note that items
# are logged in the following order, even if the actual enable lines in this file are re-ordered.
# Items that are not enabled are skipped.  See the user manual for an explanation of each register.  No log file is generated if all LOG items are commented out.
# Values are logged as base-10 decimal numbers separated by a space character.
#   ID, STATUS, BUCK, VS, IS, VB, IB, IC, IT, ET, VM, TH, BULKV, FLOATV, PWROFFV, PWRONV, WDEN, WDCNT
#
# ID Register
#LOG=ID
#
# STATUS Register
LOG=STATUS
#
# BUCK Status Register
LOG=BUCK
#
# Solar Panel Voltage in mV (VS)
LOG=VS
#
# Solar Panel Current in mA (IS)
LOG=IS
#
# Battery Voltage in mV (VB)
LOG=VB
#
# Battery Load Current in mA (IB)
LOG=IB
#
# Battery Charge Current in mA (IC)
LOG=IC
#
# Internal Temperature Sensor temperature in degrees C x 10 (IT) - (25C will show up as 250)
LOG=IT
#
# External Temperature Sensor temperature in degrees C x 10 (ET)
LOG=ET
#
# Current calculated solar panel maximum power point voltage in mV (VM)
LOG=VM
#
# Current temperature compensated battery charge voltage in mV (TH)
LOG=TH
#
# Battery BULK/ABSORPTION charge voltage threshold for 25C in mV (BULKV)
#LOG=BULKV
#
# Battery FLOAT charge voltage threshold for 25C in mV (FLOATV)
#LOG=FLOATV
#
# Low Battery power-off threshold in mV (PWROFFV)
#LOG=PWROFFV
#
# Battery power-on restart threshold in mV (PWRONV)
#LOG=PWRONV
#
# Watchdog enable Register (0=disabled, 1=enabled)(WDEN)
#LOG=WDEN
#
# Watchdog timeout count Register in seconds (WDCNT)
#LOG=WDCNT

# Charger parameters.  Uncomment the following lines to configure non-default charger configuration
# parameters.
#
# Battery BULK/ABSORPTION charge voltage threshold for 25C in mV - must be within 14000 - 15000 mV (BULKV)
#BULKV=14700
#
# Battery FLOAT charge voltage threshold for 25C in mV - must be within 13000 - 14000 mV (FLOATV)
#FLOATV=13650
#
# Low Battery power-off threshold in mV - must be within 11000 - 12000 mV (PWROFF)
#PWROFFV=11500
#
# Battery power-on restart threshold in mV - must be within 12000 - 13000 mV (PWRONV)
#PWRONV=12500

# Watchdog Enable.  Uncomment the following line to enable the daemon to control the charger's
# watchdog mechanism.  It will enable the watchdog functionality and make sure that it is
# updated in order to not cycle power.  The daemon will catch most terminal signals and
# automatically disable the watchdog before exiting.  However a hard kill of the daemon will
# leave the watchdog running and unless it is disabled via another facility, the charger will
# power cycle the system within 120 - 180 seconds.
#WATCHDOG=1



examples/motioneyeos/mpptChgDconfig.txt
# mpptChgD configuration file
#
# Configure the following optional functions:
#   1. Auto shutdown enable
#   2. TCP Access
#   3. Charger register value Logging
#   4. Charger configuration parameters
#   5. Watchdog enable
#
# All options shown below. Uncomment to enable. Configuration items have the form
# <ITEM>=<VALUE> where <VALUE> may be an integer or string value.  Items that are
# used to enable functionality (such as SHUTDOWN below) may either be commented out
# or set to 0 to disable.  Uncomment them and set the value to 1 to enable.
#
# Comments are indicated by a '#' or ';' character in the first position.
#

# ==========================================================================================
#
# Automatic Low Battery shutdown.  Uncomment the following line to enable a controlled
# OS shutdown upon detection of an imminent power down due to low battery.
SHUTDOWN=1


# ==========================================================================================
#
# Remote TCP access.  Uncomment and set the TCP_PORT to a non-zero number to enable
# TCP access to the daemon.  Remember that this may introduce security risks to your
# system.
TCP_PORT=23000
# Maximum number of simultaneous TCP connections (default is 1 connection, maximum 4)
TCP_MAX=1


# ==========================================================================================
#
# Logging parameters
#
# Configure the logfile
#LOG_FILE=/data/log/mpptChgDlog.txt
#
# Number of seconds between log entries
#LOG_DELAY=60
#
# Log items.  Uncomment charger register values you wish to log.  A comment is written
# to the log file indicating the items being logged and their order.  Items that are not
# enabled are skipped.  See the user manual for an explanation of each register.  Values
# are logged as base-10 decimal numbers separated by a space character.
#
# ID Register
#LOG=ID
#
# STATUS Register
#LOG=STATUS
#
# BUCK Status Register
#LOG=BUCK
#
# Solar Panel Voltage in mV (VS)
#LOG=VS
#
# Solar Panel Current in mA (IS)
#LOG=IS
#
# Battery Voltage in mV (VB)
#LOG=VB
#
# Battery Load Current in mA (IB)
#LOG=IB
#
# Battery Charge Current in mA (IC)
#LOG=IC
#
# Internal Temperature Sensor temperature in degrees C x 10 (IT) - (25C will show up as 250)
#LOG=IT
#
# External Temperature Sensor temperature in degrees C x 10 (ET)
#LOG=ET
#
# Current calculated solar panel maximum power point voltage in mV (VM)
#LOG=VM
#
# Current temperature compensated battery charge voltage in mV (TH)
#LOG=TH
#
# Battery BULK/ABSORPTION charge voltage threshold for 25C in mV (BULKV)
#LOG=BULKV
#
# Battery FLOAT charge voltage threshold for 25C in mV (FLOATV)
#LOG=FLOATV
#
# Low Battery power-off threshold in mV (PWROFFV)
#LOG=PWROFFV
#
# Battery power-on restart threshold in mV (PWRONV)
#LOG=PWRONV
#
# Watchdog enable Register (0=disabled, 1=enabled)(WDEN)
#LOG=WDEN
#
# Watchdog timeout count Register in seconds (WDCNT)
#LOG=WDCNT


# ==========================================================================================
#
# Charger parameters.  Uncomment the following lines to configure non-default charger
# configuration parameters.
#
# Battery BULK/ABSORPTION charge voltage threshold for 25C in mV (BULKV)
#  Must be within the range 14000 - 15000 mV
#BULKV=14700
#
# Battery FLOAT charge voltage threshold for 25C in mV (FLOATV)
#  Must be within the range 13000 - 14000 mV
#FLOATV=13650
#
# Low Battery power-off threshold in mV  (PWROFFV)
#  Must be within the range 11000 - 12000 mV
#PWROFFV=11500
#
# Battery power-on restart threshold in mV  (PWRONV)
#  Must be within the range 12000 - 13000 mV
#PWRONV=12500


# ==========================================================================================
#
# Watchdog Enable.  Uncomment the following line to enable the daemon to control the charger's
# watchdog mechanism.  It will enable the watchdog functionality and make sure that it is
# updated in order to not cycle power.  The daemon will catch most terminal signals and
# automatically disable the watchdog before exiting.  However a hard kill of the daemon will
# leave the watchdog running and unless it is disabled via another facility, the charger will
# power cycle the system within 120 - 180 seconds.
#WATCHDOG=1



firmware/SolarMpptCharger/inc/config.h
/*
 * config.h
 *
 *  Global default system configuration parameters
 *
 * Copyright (c) 2018-2019 danjuliodesigns, LLC.  All rights reserved.
 *
 * SolarMpptCharger is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * SolarMpptCharger is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * See <http://www.gnu.org/licenses/>.
 *
 */

#ifndef INC_CONFIG_H_
#define INC_CONFIG_H_

// Firmware version
//  Each field should be limited to 4-bits
#define FW_ID               1
#define FW_VER_MAJOR        1
#define FW_VER_MINOR        2

// Charge Voltage Thresholds (mV)
#define V_NIGHT_THRESH      3500
#define V_MAX_SOLARV        21000
#define V_MIN_GOOD_SOLAR    18000
#define V_MIN_SOLARV        12000
#define V_MIN_IDLE_2_FLT    12700
#define V_BULK_DEFAULT      14700
#define V_FLOAT_DEFAULT     13650
#define V_DELTA_CHANGE      20

// Load Control Voltage Thresholds (mV)
#define V_LOAD_OFF          11500
#define V_LOAD_ON           12500

// Bad Battery Voltage Threshold (mV)
#define V_BAD_BATTERY       10500

// Adjustable threshold limits
#define V_BULK_MIN          14000
#define V_BULK_MAX          15000
#define V_FLOAT_MIN         13000
#define V_FLOAT_MAX         14000
#define V_LOAD_OFF_MIN      11000
// V_LOAD_OFF_MAX set by current V_LOAD_ON
#define V_LOAD_ON_MIN       12000
#define V_LOAD_ON_MAX       13000

// Hysteresis to prevent night timer operation toggling around a critically low
// battery voltage
#define PWR_LB_HYST_MV      250

// Current Thresholds (mA)
#define I_SOLAR_MIN_THRESH  10
#define I_SOLAR_MAX         2000
#define I_ABS_CUTOFF        300

// Power Thresholds (mW)
#define P_MIN_THRESH        100

// Charging temperature limits (C)
#define TEMP_LIMIT_HIGH     50
#define TEMP_LIMIT_LOW     -20
#define TEMP_LIMIT_HYST     5

// Temperature compensation (mV for 6 cells per degree C)
//  x 10 to enable rounding of an integer calculation
#define V_BULK_COMP_X10    -300
#define V_FLOAT_COMP_X10   -198

// Buck regulator voltage regulation hysteresis (mV)
#define V_BUCK_HYST         15

// Timeouts (seconds) - should fit in an uint16_t
#define WAKE_TIMEOUT          60
#define NIGHT_TIMEOUT         300
#define MPPT_SCAN_TIMEOUT     600
#define HIGH_CHARGE_TIMEOUT   36000
#define CHG_RCVR_PERIOD       3
#define LOWPWR_TIMEOUT        60
#define PWROFF_WARN_TIMEOUT   60
#define PWROFF_DEF_WD_TIMEOUT 10
#define PWROFF_LB_CHG_TIMEOUT 3600
// Timeouts (seconds) - should fit in an uint8_t
#define LOW_PROD_TIMEOUT      15
#define ABS_TERM_TIMEOUT      30

// MPPT Step voltages
#define MPPT_HI_I_STEP_MV   50
#define MPPT_MID_I_STEP_MV  100
#define MPPT_LO_I_STEP_MV   200
#define MPPT_HI_I_STEP_MA   200
#define MPPT_LO_I_STEP_MA   100

// MPPT Scan Step voltage
#define MPPT_SCAN_STEP_MV   200

// Diagnostic output port bits for use by any module
SI_SBIT(DIAG1, SFR_P1, 6);
SI_SBIT(DIAG2, SFR_P1, 7);


#endif /* INC_CONFIG_H_ */
